File: src/main/java/caelum/focalworks/registry/FocalworksActions.java
package caelum.focalworks.registry;


import at.petrak.hexcasting.api.casting.ActionRegistryEntry;
import at.petrak.hexcasting.api.casting.castables.Action;
import at.petrak.hexcasting.api.casting.math.HexDir;
import at.petrak.hexcasting.api.casting.math.HexPattern;
import net.minecraft.core.Registry;
import at.petrak.hexcasting.common.lib.hex.HexActions;
import caelum.focalworks.casting.actions.*;
import static caelum.focalworks.Focalworks.id;


public class FocalworksActions {
    public static void init() {
        register("rig_read","wwaqwqqqwqwqqwwqqeqqwwqq",HexDir.EAST,new OpRigRead());
        register("rig_write","wwdeweeeweweewweeqeewwee",HexDir.EAST,new OpRigWrite());
        register("crack_rig","wwaqwqqqwqedwewwewdawdwwwwdw",HexDir.EAST,new OpRigWrite());
        register("flip_spellbook","wwaqwqqqwqeawqwwqwadaeqqea",HexDir.EAST,new OpCrackRig());
    }
    private static ActionRegistryEntry register(
            String name,
            String signature,
            HexDir startDir,
            Action action
    ) {
        return Registry.register(HexActions.REGISTRY, id(name), new ActionRegistryEntry(HexPattern.fromAngles(signature, startDir), action));
    }
}

File: common/src/main/kotlin/io/github/c_caelum/focalworks/registry/FocalworksActions.kt
package io.github.c_caelum.focalworks.registry

import at.petrak.hexcasting.api.casting.ActionRegistryEntry
import at.petrak.hexcasting.api.casting.castables.Action
import at.petrak.hexcasting.api.casting.math.HexDir
import at.petrak.hexcasting.api.casting.math.HexPattern
import at.petrak.hexcasting.common.lib.HexRegistries
import at.petrak.hexcasting.common.lib.hex.HexActions
import io.github.c_caelum.focalworks.casting.actions.*
import io.github.c_caelum.focalworks.casting.actions.spells.*

object FocalworksActions : FocalworksRegistrar<ActionRegistryEntry>(
    HexRegistries.ACTION,
    { HexActions.REGISTRY },
) {
    val RIG_READ = make("rig_read",HexDir.NORTH_WEST,"qaqwwaqqqqqwede", OpRigRead)

    private fun make(name: String, startDir: HexDir, signature: String, action: Action) =
        make(name, startDir, signature) { action }

    private fun make(name: String, startDir: HexDir, signature: String, getAction: () -> Action) = register(name) {
        ActionRegistryEntry(HexPattern.fromAngles(signature, startDir), getAction())
    }
}
